librz/crypto/p/crypto_sm4_ecb.c

```diff
@@ -0,0 +1,234 @@
+	rz_write_at_be32(buf, val, 0);
+	for (int i = 0; i < 4; i++) {
+		buf[i] = sbox[buf[i]];
+	}
+	return rz_read_be32(buf);
+}
+
+static ut32 sm4_ecb_get_round_key(ut32 val) {
+	ut32 x = sm4_ecb_transform_internal(val);
+	return x ^ _rol32(x, 13) ^ _rol32(x, 23);
+}
+
+static ut32 sm4_ecb_transform(ut32 val) {
+	ut32 x = sm4_ecb_transform_internal(val);
+	return x ^ _rol32(x, 2) ^ _rol32(x, 10) ^ _rol32(x, 18) ^ _rol32(x, 24);
+}
+
+static void sm4_setkey_internal(sm4_state *s, const ut8 *key) {
+	ut32 mk[4];
+	ut32 rk[4];
+	ut32 i, temp;
+
+	for (i = 0; i < 4; i++) {
+		mk[i] = rz_read_be32(key + 4 * i);
+	}
+	for (i = 0; i < 4; i++) {
+		rk[i] = mk[i] ^ FK[i];
+	}
+
+	for (i = 0; i < 32; i++) {
+		temp = rk[0] ^ sm4_ecb_get_round_key(rk[1] ^ rk[2] ^ rk[3] ^ CK[i]);
+		s->round_keys[i] = temp;
+		rk[0] = rk[1];
+		rk[1] = rk[2];
+		rk[2] = rk[3];
+		rk[3] = temp;
+	}
+}
+
+static void sm4_setkey_enc(sm4_state *s, const ut8 *key) {
+	sm4_setkey_internal(s, key);
+}
+static void sm4_setkey_dec(sm4_state *s, const ut8 *key) {
+	sm4_setkey_internal(s, key);
+	for (int i = 0; i < 16; i++) {
+		ut32 temp = s->round_keys[i];
+		s->round_keys[i] = s->round_keys[31 - i];
+		s->round_keys[31 - i] = temp;
+	}
+}
+
+static bool sm4_ecb_set_key(RzCrypto *cry, const ut8 *key, int keylen, int mode, int direction) {
+	rz_return_val_if_fail(cry->user && key, false);
+	sm4_state *s = (sm4_state *)cry->user;
+	cry->dir = direction;
+	if (keylen != SM4_KEY_SIZE) {
+		return false;
+	}
+	if (direction == RZ_CRYPTO_DIR_ENCRYPT) {
+		sm4_setkey_enc(s, key);
+	} else {
+		sm4_setkey_dec(s, key);
+	}
+	return true;
+}
+
+static int sm4_ecb_get_key_size(RzCrypto *cry) {
+	return SM4_KEY_SIZE;
+}
+
+static bool sm4_use(const char *algo) {
+	return !strcmp(algo, "sm4-ecb");
+}
+
+static void sm4_round(const ut32 *round_key, const ut8 *input, ut8 *output) {
+	ut32 x[36], i;
+	for (i = 0; i < 4; i++) {
+		x[i] = rz_read_be32(input + 4 * i);
+	}
+	for (i = 0; i < 32; i++) {
+		x[i + 4] = x[i] ^ sm4_ecb_transform(x[i + 1] ^ x[i + 2] ^ x[i + 3] ^ round_key[i]);
+	}
+
+	rz_write_at_be32(output, x[35], 0);
+	rz_write_at_be32(output, x[34], 4);
+	rz_write_at_be32(output, x[33], 8);
+	rz_write_at_be32(output, x[32], 12);
+}
+static void sm4_ecb_encrypt(const ut32 *round_key, int length, const ut8 *input, ut8 *output) {
+	int times = length / 16;
+	for (int i = 0; i < times; i++) {
+		sm4_round(round_key, input, output);
+		input += 16;
+		output += 16;
+	}
+}
+
+static int get_next_available_len(int curr_len) {
+	return ((curr_len + 15) / 16) * 16;
+}
+
+static bool update(RzCrypto *cry, const ut8 *buf, int len) {
+	rz_return_val_if_fail(cry, 0);
+	sm4_state *state = (sm4_state *)cry->user;
+	int old_len = len;
+	if (len < 1) {
+		return false;
+	}
+	if (cry->dir == RZ_CRYPTO_DIR_ENCRYPT) {
+		if (len % 16 == 0) {
+			len += 16;
+		} else {
+			len = get_next_available_len(len);
+		}
+	}
+	ut8 *output = (ut8 *)calloc(1, len);
+	ut8 *padded_input = (ut8 *)calloc(1, len);
```

Comment:
check for failed allocs.

Reply:
In various cases, the input must be padded with null bytes. Therefore, I used calloc to initialize the allocated memory with null bytes.

Reply:
ok, makes sense. still check for the allocs if it fails.